### PR TITLE
Add warp Player to map position

### PR DESCRIPTION
**Pull request by Patrick2562**
request #202 

image

image

### CODE REVIEW

File: [admin]/admin/client/gui/admin_warp.lua [hunk 1]
```diff
@@ -46,26 +48,115 @@ function aPlayerWarpClose ( destroy )
 	end
 end
 
+
+function aPlayerWarpToPosition ( )
+	if ( aWarpToPositionForm == nil ) then
+		local x, y = guiGetScreenSize()
+		local h	   = y * 0.75
+		aWarpToPositionForm		= guiCreateWindow ( x / 2 - h / 2, y / 2 - h / 2, h, h + 40, "Player Warp To Position", false )
+		aWarpToPositionMap		= guiCreateStaticImage ( 10, 25, h - 20, h - 20, "client/images/map.png", false, aWarpToPositionForm )
```
Comment (Dezash) [inline [admin]/admin/client/gui/admin_warp.lua:57]: Seems like you forgot to add the image to your PR

Comment (galeapatrik) [reply [admin]/admin/client/gui/admin_warp.lua:57]: Ohh boi... nice find xd

**Comment by iiN3xT** [comment]
I think this should be on admin2, no?

**Comment by galeapatrik** [comment]
> I think this should be on admin2, no?
It's already done, but I can do it in admin2 too.

**Comment by iiN3xT** [comment]
> > It's already done, but I can do it in admin2 too.

That would be nice since we're trying to give players a reason to move to admin2

**Comment by Abasalt-yar** [comment]
> > I think this should be on admin2, no?
> 
> It's already done, but I can do it in admin2 too.

It's Good to be on admin2  | Thanks :smiley:

**Comment by qaisjp** [comment]
Tabs are introduced in this PR where the rest of the file uses spaces

File: [admin]/admin/client/gui/admin_warp.lua [hunk 2]
```diff
@@ -46,26 +48,115 @@ function aPlayerWarpClose ( destroy )
...
+	guiSetText ( aWarpToPositionX, tpX ); guiSetText ( aWarpToPositionY, tpY ); guiSetText ( aWarpToPositionZ, tpZ )
+	return tpX, tpY, tpY
+end
+
+local function getTeleportPosition ( )
+	return tonumber ( guiGetText ( aWarpToPositionX ) ) or 0, tonumber ( guiGetText ( aWarpToPositionY ) ) or 0, guiGetText ( aWarpToPositionZ )
+end
+
+local function warpPlayerToPositionTrigger ( )
+	local function warpToPosition ( player, x, y, z )
```
Comment (Dezash) [inline [admin]/admin/client/gui/admin_warp.lua:102]: This function doesn't have to be nested

File: [admin]/admin/client/gui/admin_warp.lua [hunk 3]
```diff
@@ -46,26 +48,115 @@ function aPlayerWarpClose ( destroy )
...
+		local distance = getElementDistanceFromCentreOfMassToBaseOfModel ( player )
+		triggerServerEvent ( "aPlayer", getLocalPlayer(), aWarpSelectPointer, "warptoposition", { x, y, z + distance + 0.25 } )
```
Comment (Dezash) [inline [admin]/admin/client/gui/admin_warp.lua:104]:
```suggestion
		triggerServerEvent ( "aPlayer", localPlayer, player, "warptoposition", { x, y, z + distance + 0.25 } )

```

File: [admin]/admin/client/gui/admin_warp.lua [hunk 4]
```diff
@@ -46,26 +48,115 @@ function aPlayerWarpClose ( destroy )
...
+		fadeCamera ( false, 0 )
+		setElementFrozen ( getPedOccupiedVehicle ( localPlayer ) or localPlayer, true )
+		setCameraMatrix ( x, y, 0 )
+		setTimer ( function ( )
+			local hit, _, _, hitZ = processLineOfSight(x, y, 3000, x, y, -3000)
+			setCameraTarget ( localPlayer )
+			setElementFrozen ( getPedOccupiedVehicle ( localPlayer ) or localPlayer, false )
+			fadeCamera ( true, 0.3 )
+			if not hit then return end
+			warpToPosition ( aWarpSelectPointer, x, y, hitZ )
```
Comment (Dezash) [inline [admin]/admin/client/gui/admin_warp.lua:120]: Could aWarpSelectPointer point to a player who just disconnected? Perhaps an isElement check should be done at the beginning of the function.

File: [admin]/admin/client/gui/admin_warp.lua [hunk 5]
```diff
@@ -46,26 +48,115 @@ function aPlayerWarpClose ( destroy )
...
+	else
+		warpToPosition ( aWarpSelectPointer, x, y, z )
+	end
+end
+
 function aClientWarpDoubleClick ( button )
 	if ( button == "left" ) then
 		if ( source == aWarpList ) then
 			if ( guiGridListGetSelectedItem ( aWarpList ) ~= -1 ) then
 				triggerServerEvent ( "aPlayer", getLocalPlayer(), aWarpSelectPointer, "warpto", getPlayerFromNick ( guiGridListGetItemPlayerName ( aWarpList, guiGridListGetSelectedItem ( aWarpList ), 1 ) ) )
```
Comment (Dezash) [inline [admin]/admin/client/gui/admin_warp.lua:131]:
```suggestion
				triggerServerEvent ( "aPlayer", localPlayer, aWarpSelectPointer, "warpto", getPlayerFromNick ( guiGridListGetItemPlayerName ( aWarpList, guiGridListGetSelectedItem ( aWarpList ), 1 ) ) )

```

File: [admin]/admin/client/gui/admin_warp.lua [hunk 6]
```diff
@@ -46,26 +48,115 @@ function aPlayerWarpClose ( destroy )
...
 	end
 end
 
-function aClientWarpClick ( button )
+function aClientWarpClick ( button, state, absX, absY )
 	if ( button == "left" ) then
+		-- Player Warp Management
 		if ( source == aWarpSelect ) then
 			if ( guiGridListGetSelectedItem ( aWarpList ) ~= -1 ) then
 				triggerServerEvent ( "aPlayer", getLocalPlayer(), aWarpSelectPointer, "warpto", getPlayerFromNick ( guiGridListGetItemPlayerName ( aWarpList, guiGridListGetSelectedItem ( aWarpList ), 1 ) ) )
```
Comment (Dezash) [inline [admin]/admin/client/gui/admin_warp.lua:145]:
```suggestion
				triggerServerEvent ( "aPlayer", localPlayer, aWarpSelectPointer, "warpto", getPlayerFromNick ( guiGridListGetItemPlayerName ( aWarpList, guiGridListGetSelectedItem ( aWarpList ), 1 ) ) )

```

File: [admin]/admin2/client/widgets/admin_warp.lua [hunk 7]
```diff
@@ -47,6 +50,88 @@ function aPlayerWarpClose(destroy)
...
+local function getTeleportPosition ( )
+    return tonumber(guiGetText(aWarpToPositionX)) or 0, tonumber(guiGetText(aWarpToPositionY)) or 0, guiGetText(aWarpToPositionZ)
+end
+
+local function warpPlayerToPositionTrigger()
+    local function warpToPosition(player, x, y, z)
+        local distance = getElementDistanceFromCentreOfMassToBaseOfModel(player)
+        triggerServerEvent(
+            "aPlayer",
+            getLocalPlayer(),
```
Comment (Dezash) [inline [admin]/admin2/client/widgets/admin_warp.lua:108]:
```suggestion
            localPlayer,
```

**Comment by galeapatrik** [comment]
> Using "warptoposition" as action requires "command.warptoposition" permission, which is not present in default ACL configuration. I'd suggest using "warpto" and determine whether it is a map warp by inspecting _data_ argument.
> So if _data_ is of type player, warpPlayer is called and if _data_ is of type table then it's a map warp.

Check this, I modified the warpPlayer function, or would I do a new one? (warpToPosition or something)

**Comment by Dezash** [comment]
> Check this, I modified the warpPlayer function, or would I do a new one? (warpToPosition or something)

I think that's fine, but _warp_ function should probably not be nested inside _warpPlayer_.

**Comment by galeapatrik** [comment]
> > Check this, I modified the warpPlayer function, or would I do a new one? (warpToPosition or something)
> 
> I think that's fine, but _warp_ function should probably not be nested inside _warpPlayer_.

It was nested by default, I didn't do it.

**Comment by Dezash** [comment]
> It was nested by default, I didn't do it.

Yes, I know, but it would be good to improve the old code since you're working on it anyway.

**Comment by galeapatrik** [comment]
> Yes, I know, but it would be good to improve the old code since you're working on it anyway.

I'm gonna move it outside.

**Comment by Dezash** [comment]
I've been running the admin resource with these changes for two weeks now and there were no issues so far 👍

**Comment by jlillis** [comment]
This PR currently contains merge conflicts. If you can resolve them, I'd be happy to review and merge it.

**Comment by Patrick2562** [comment]
> This PR currently contains merge conflicts. If you can resolve them, I'd be happy to review and merge it.

Resolved.

**Review by jlillis — APPROVED** [review]
Works as advertised. There's some cleanup that may need to be done to admin2 in the future but I'd rather not hold this change back right now.

**Pull request merged**